pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java

```diff
@@ -240,13 +280,48 @@ public void addFailed(ManagedLedgerException exception, Object ctx) {
                 if (exception instanceof ManagedLedgerException.ManagedLedgerAlreadyClosedException) {
                     managedLedger.readyToCreateNewLedger();
                 }
-                buf.release();
                 completableFuture.completeExceptionally(new PersistenceException(exception));
             }
         }, null);
         return completableFuture;
     }
 
+    /**
+     * Build the index mapping of Transaction pending ack log (aka t-log) and Topic message log (aka m-log).
+     * When m-log has been ack, t-log which holds m-log is no longer useful, this method builder the mapping of them.
+     *
+     * If a Ledger Entry has many t-log, we only need to care about the record that carries the largest acknowledgement
+     * info. Because all Commit/Abort log after this record describes behavior acknowledgement, if the behavior
+     * acknowledgement has been handle correct, these Commit/Abort log is no longer useful.
+     * @param logPosition The position of batch log Entry.
+     * @param logList Pending ack log records in a batch log Entry.
+     */
+    private void handleMetadataEntry(PositionImpl logPosition, List<PendingAckMetadataEntry> logList) {
+        // Find the record that carries the largest ack info, and call "handleMetadataEntry(position, pendingAckLog)"
+        PendingAckMetadataEntry pendingAckLogHasMaxAckPosition = null;
+        PositionImpl maxAcknowledgementPosition = this.maxAckPosition;
+        for (int i = logList.size() - 1; i >= 0; i--){
+            PendingAckMetadataEntry pendingAckLog = logList.get(i);
+            if (pendingAckLog.getPendingAckOp() == PendingAckOp.ABORT
+                    && pendingAckLog.getPendingAckOp() == PendingAckOp.COMMIT) {
+                continue;
+            }
+            if (pendingAckLog.getPendingAckMetadatasList().isEmpty()){
+                continue;
+            }
+            for (PendingAckMetadata ack : pendingAckLog.getPendingAckMetadatasList()){
+                if (maxAcknowledgementPosition.compareTo(ack.getLedgerId(), ack.getEntryId()) < 0){
+                    maxAcknowledgementPosition = PositionImpl.get(ack.getLedgerId(), ack.getEntryId());
+                    pendingAckLogHasMaxAckPosition = pendingAckLog;
+                }
+            }
+        }
```

Comment:
only the logList.size() is batchSize, we handle once is enough. don't every add ops handle